msg_queue.go

```diff
@@ -38,44 +63,8 @@ func (m *msgQueue) readMessageWithDiscards(st State, current *View) (*MessageReq
 	m.queueLock.Lock()
 	defer m.queueLock.Unlock()
 
-	discarded := []*MessageReq{}
 	queue := m.getQueue(st)
-
-	for {
-		if queue.Len() == 0 {
-			return nil, discarded
-		}
-		msg := queue.head()
-
-		// check if the message is from the future
-		if st == RoundChangeState {
-			// if we are in RoundChangeState we only care about sequence
-			// since we are interested in knowing all the possible rounds
-			if msg.View.Sequence > current.Sequence {
-				// future message
-				return nil, discarded
-			}
-		} else {
-			// otherwise, we compare both sequence and round
-			if cmpView(msg.View, current) > 0 {
-				// future message
-				return nil, discarded
-			}
-		}
-
-		// at this point, 'msg' is good or old, in either case
-		// we have to remove it from the queue
-		heap.Pop(queue)
-
-		if cmpView(msg.View, current) < 0 {
-			// old value, try again
-			discarded = append(discarded, msg)
-			continue
-		}
-
-		// good value, return it
-		return msg, discarded
-	}
+	return queue.readMessageWithDiscardsLocked(current, st)
```

Comment:
Is there any benefit on moving this logic out?

Reply:
Yes, reusing its logic and not having it at two places, since we are using it to read from both `pendingCommitMsgsQueue` (which is not related to any specific PBFT state) and from other queues (related to PBFT states, such as `validateStateQueue`, `roundChangeStatteQueue` etc.).

I was thinking initially of reusing `validateStateQueue` for pending commit messages as well, but that would require expanding commit messages with additional flag indicating its state (pending/non-validated vs confirmed/validated). Documented [here](https://www.notion.so/RFC-28-0e69004ef82241af88b16e41d51a15c2#86e6cc3ffd4444919d34087d78e4d8be).

Reply:
This is going to be a problem later on. This function is not only doing the discard of messages but doing internal logic (i.e RoundChangeState condition). You do not need that kind of logic when dealing with pending commit messages.

Reply:
> This is going to be a problem later on.

I don't see this part of the code is going to be changed ever.

> You do not need that kind of logic when dealing with pending commit messages.

Ok agree with that part. In that case, I'd suggest introducing additional parameter which should handle message differently, based on given `PbftState` value.

However, please don't let we find ourselves forking logic and copying almost same functions multiple times, just because they aren't 100% the same.